### PR TITLE
NIFI-10234 Implement PutIoTDB

### PR DESCRIPTION
**Pull request by xuanronaldo**
### Properties of PutIoTDB

| property      | description                                                  | default value | necessary |
| ------------- | ------------------------------------------------------------ | ------------- | --------- |
| Host          | The host of IoTDB.                                           | null          | true      |
| Port          | The port of IoTDB.                                           | 6667          | true      |
| Username      | Username to access the IoTDB.                                | null          | true      |
| Password      | Password to access the IoTDB.                                | null          | true      |
| Record Reader | Specifies the type of Record Reader controller service to use <br />for parsing the incoming data and determining the schema. | null          | true      |
| Schema        | The schema that IoTDB needs doesn't support good by NiFi.<br/>Therefore, you can define the schema here. <br />Besides, you can set encoding type and compression type by this method.<br />If you don't set this property, the inferred schema will be used.<br /> It can be updated by expression language. | null          | false     |
| Aligned       | Whether using aligned interface?  It can be updated by expression language. | false         | false     |
| MaxRowNumber  | Specifies the max row number of each tablet.  It can be updated by expression language. | 1024          | false     |

### Inferred Schema of Flowfile

There are a couple of rules about flowfile:

1. The flowfile can be read by `Record Reader`.
2. The schema of flowfile must contains a field `Time`, and it must be the first.
3. The data type of time must be `STRING` or `LONG`.
4. Fields excepted time must start with `root.`.
5. The supported data types are `INT`, `LONG`, `FLOAT`, `DOUBLE`, `BOOLEAN`, `TEXT`.

### Convert Schema by property

As mentioned above, converting schema by property which is more flexible and stronger than inferred schema. 

The structure of property `Schema`:

```json
{
	"timeType": "long",
	"field": [{
		"tsName": "root.sg.d1.s1",
		"dataType": "INT32",
		"encoding": "RLE",
		"compressionType": "GZIP"
	}, {
		"tsName": "root.sg.d1.s2",
		"dataType": "INT64",
		"encoding": "RLE",
		"compressionType": "GZIP"
	}]
}
```

**Note**

1. The first column must be `Time`. The rest must be arranged in the same order as in `field` of JSON.
1. The JSON of schema must contain `timeType` and `fields`.
2. There are only two options `long` and `string` for `timeType`.
3. The columns `tsName` and `dataType` must be set.
4. The tsName must start with `root.`.
5. The supported `dataTypes` are `INT32`, `INT64`, `FLOAT`, `DOUBLE`, `BOOLEAN`, `TEXT`.
6. The supported `encoding` are `PLAIN`, `DICTIONARY`, `RLE`, `DIFF`, `TS_2DIFF`, `BITMAP`, `GORILLA_V1`, `REGULAR`, `GORILLA`.
7. The supported `compressionType` are `UNCOMPRESSED`, `SNAPPY`, `GZIP`, `LZO`, `SDT`, `PAA`, `PLA`, `LZ4`.

## Relationships

| relationship | description                                          |
| ------------ | ---------------------------------------------------- |
| success      | Data can be written correctly or flow file is empty. |
| failure      | The shema or flow file is abnormal.                  |

### CODE REVIEW

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the contribution @xuanronaldo! Supporting integration with Apache IoTDB through the standard Java API looks like a very useful feature.

Unfortunately there several general issues that need to be resolved before a more thorough implementation review can go forward.

1. The `Static Analysis` check indicates several style problems. At a quick glance, some of the problems appear to be related to the use of asterisk imports. Please run a Maven build with the `contrib-check` profile to evaluate the issues
2. All `maven.compiler` properties must be removed. Although NiFi will be deprecating support for Java 8 soon, it is still a supported version, and that property cannot be overridden in specific modules
3. All packages must begin with `org.apache.nifi`, instead of `org.apache.iotdb`
4. All Maven `groupId` elements must contain `org.apache.nifi`, instead of `org.apache.iotdb`

There are a couple other general recommendations worth noting:

1. The Jackson library should be used for JSON processing unless there is a compelling reason to use a different library
2. Use of inline anonymous class declarations for collections such as HashMap, HashSet, and ArrayList should be avoided. Although this has some convenience value, it is not recommended in either runtime code or test code.

Thanks again for working on this new feature. After addressing these issues, the changes should be in a good position for a more detailed review.

**Comment by exceptionfactory** [comment]
One additional note, Apache NiFi just released version 1.17.0, so the current main branch is now on 1.18.0-SNAPSHOT. Please rebase and update to the current version.

Thanks!

**Comment by xuanronaldo** [comment]
> 

Ok, I got it. Thanks for your suggestions. I'm working on it.

**Comment by xuanronaldo** [comment]
I have pushed the code, and all checks have passed. What should I do next to merge the branch?

**Comment by exceptionfactory** [comment]
Thanks for making the updates @xuanronaldo. Merging requires approval from at least one committer. I will try to take a look at the latest version soon.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:94]:
```suggestion
                    .allowableValues("true", "false")
```

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:93]:
```suggestion
                    .description("Whether to use the Apache IoTDB Aligned Timeseries interface")
```

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:105]:
```suggestion
                            "Specifies the max row number of each Apache IoTDB Tablet")
```

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:107]: The default value should be removed if this property is not required.
```suggestion
```

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:141]: These `getProperty()` calls need to include `evaluateAttributeExpressions(flowFile)` in order to resolve FlowFile-based attributes.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:152]: This null check should be moved right after `processSession.get()`, instead of inside this try-catch block.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:146]:
```suggestion
        final boolean aligned = alignedProperty != null ? Boolean.valueOf(alignedProperty) : false;
```

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:183]: `ObjectMapper` should be declared as a static member and reused.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:200]: This message should include some additional details, such as the FlowFile and record number:
```suggestion
                        getLogger().error("{} Record [{}] time format not supported", flowFile, recordNumber);
```

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:236]: It looks like this exception should be logged, at least at the debug level.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:272]: The exception should be logged with a message to ensure the full stack trace is available:
```suggestion
            getLogger().error("Processing failed {}", flowFile, e);
```

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:242]: It would be helpful to refactor this method into smaller component parts to avoid deeply-nested loops.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/model/Schema.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/model/Schema.java:32]: It would be helpful to name this class more specifically given the various types of schemas in the context of Apache NiFi. What do you think of `DatabaseSchema` or `TimeDatabaseSchema`?

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/test/java/org/apache/nifi/processors/AbstractIoTDBUT.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/test/java/org/apache/nifi/processors/AbstractIoTDBUT.java:30]: The JUnit 5 `Assertions` static methods should be used instead of JUnit 4.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/test/java/org/apache/nifi/processors/PutIoTDBIT.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/test/java/org/apache/nifi/processors/PutIoTDBIT.java:32]: Should be changed to JUnit 5 `Assertions` and it would be better to use static imports for assert methods.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:179]: All log messages should have a readable message, in addition to passing the relevant value.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:172]: Closing these objects should be unnecessary with the use of try-with-resources.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:215]: The purpose of this variable is somewhat unclear based on the name.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:159]: This variable, and perhaps others, should be declared and assigned in the same location for better readability.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for making the initial changes @xuanronaldo.

I took a closer look at the implementation and noted a number of areas for improvement. The division of the abstract class and concrete class is helpful to indicate the core functionality. The core logic in `onTrigger` was difficult to follow at points because of the layout and nesting, so some general refactoring would be helpful.

**Comment by xuanronaldo** [comment]
> Thanks for making the initial changes @xuanronaldo.
> 
> I took a closer look at the implementation and noted a number of areas for improvement. The division of the abstract class and concrete class is helpful to indicate the core functionality. The core logic in `onTrigger` was difficult to follow at points because of the layout and nesting, so some general refactoring would be helpful.

Thanks for your suggestions. I will update the code asap.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (xuanronaldo) [reply nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:172]: I have tried to delete these, but it will throw an exception `java.lang.IllegalStateException: FlowFile[0,16972544312800.mockFlowFile,0B] already in use for an active callback or InputStream created by ProcessSession.read(FlowFile) has not been closed`.

Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:165]: Thanks for the reply @xuanronaldo, does this occur when calling `ProcessSession.transfer()`? If you move it outside of the try-catch block, or change the `read()` signature to use a callback, that should resolve this error.

**Comment by xuanronaldo** [comment]
I have merged the main branch to resolve the problem which happened in workflow. Please approve running workflows agin. @exceptionfactory

**Comment by xuanronaldo** [comment]
All checks have passed, could you please take a look at my latest commit? @exceptionfactory

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/test/java/org/apache/nifi/processors/PutIoTDBIT.java
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/test/java/org/apache/nifi/processors/PutIoTDBIT.java:32]: @xuanronaldo Please review this comment and update the assertions to use JUnit 5 instead of JUnit 4. Thanks!

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/test/java/org/apache/nifi/processors/PutIoTDBUT.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/test/java/org/apache/nifi/processors/PutIoTDBUT.java:25]: See the other note about changing to use JUnit 5 `Assertions`.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/pom.xml
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/pom.xml:80]: This dependency is not necessary and should be removed

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:182]: The trailing slash character should be removed:
```suggestion
                        getLogger().error("{} Record [{}] time format not supported", flowFile, recordNumber");
```

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:192]: Recommend breaking this into multiple lines for better readability.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:128]: This transfer should be removed because `flowFile` is null.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:69]: Recommend naming this property `Schema Template` to align with the IoTDB terminology.

File: nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDB.java:74]: The wording of this description is not quite clear, and does not need to state what NiFi does not support. Recommend adjusting the wording to include more details about the expected format, along the following lines:

The Apache IoTDB Schema Template defined using JSON. The Processor will infer the IoTDB Schema when this property is not configured.

The [IoTDB Schema Template](https://iotdb.apache.org/UserGuide/Master/Data-Concept/Schema-Template.html) documentation provides some details, but is their an official format specification? It would be helpful to link to that documentation if available, otherwise it will be unclear how to define this property.

**Review by exceptionfactory — COMMENTED** [review]
Thanks for making the latest updates @xuanronaldo, this is looking closer to completion. I noted a few additional minor details, and I had a particular question about the expected format of the IoTDB Schema Template.

I plan to run through some additional tests soon.

**Comment by xuanronaldo** [comment]
> 

Okay, I got it. I will update the code asap. Thanks for your suggestions.

**Comment by xuanronaldo** [comment]
Hi, @exceptionfactory . Sorry about committing the code so late. I got lots of work to do, recently. Besides, I have updated a document about `IoTDB Schema Template` in our website as you requested. The link of it was added into the description of `Schema Template`.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the update @xuanronaldo. The link listed in the Apache IoTDB documentation appears to describe this processor, as opposed to a generalized IoTDB Schema Template. These seems like the Schema Template is specific to this processor property, as opposed to being a generalized Apache IoTDB standard representation. Is that correct? Perhaps this is the best way to represent that IoTDB Schema, but if there is a schema representation that is specific to IoTDB, not NiFi, that would be ideal. I will take a closer look at the implementation and documentation.

It looks like the most recent commits resulted in introducing several unrelated changes, found in the nifi-hubspot-bundle and nifi-standard-bundle. Please review and revert these changes so that the latest version of the pull request is limited to IoTDB Processor changes. Thanks!

**Comment by xuanronaldo** [comment]
Yep, you are rigth @exceptionfactory . The `template` in IoTDB is a structure to reduce the memory usage. However, the `Schema Template` here is just an adapter to define the schema in IoTDB. If you want to create schema by native way, the DDL or coding is necessary. Therefore, I designed the `Schema Template` to make it.

**Comment by xuanronaldo** [comment]
I have reverted these changes. However, several files were still marked for modification. I have no idea what to do. @exceptionfactory

**Comment by exceptionfactory** [comment]
> I have reverted these changes. However, several files were still marked for modification. I have no idea what to do. @exceptionfactory

I see the files are still marked as changed. If it is necessary to squash all current changes in order to remove the merge commits and avoid including these files marked as changed, that would be fine too. Either way, the latest version should just reflect the new components.

**Comment by xuanronaldo** [comment]
> > I have reverted these changes. However, several files were still marked for modification. I have no idea what to do. @exceptionfactory
> 
> I see the files are still marked as changed. If it is necessary to squash all current changes in order to remove the merge commits and avoid including these files marked as changed, that would be fine too. Either way, the latest version should just reflect the new components.

I have no idea about squash all current changes. Or I can create a new branch and submit another PR?

**Comment by exceptionfactory** [comment]
Thanks for following up @xuanronaldo. Sure, if it is easier for you to create a new branch with the current version, that's fine as well. It is easy to link to this PR for background, and we can work from the new PR.

**Pull request closed without merging**
